fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/KeyValueCursorBase.java

```diff
@@ -0,0 +1,410 @@
+            this.lowEndpoint = lowEndpoint;
+            return getThis();
+        }
+
+        public T setHigh(@Nullable Tuple high, @Nonnull EndpointType highEndpoint) {
+            setHigh(high != null ? subspace.pack(high) : subspace.pack(), highEndpoint);
+            return getThis();
+        }
+
+        @SpotBugsSuppressWarnings(value = "EI2", justification = "copies are expensive")
+        public T setHigh(@Nonnull byte[] highBytes, @Nonnull EndpointType highEndpoint) {
+            this.highBytes = highBytes;
+            this.highEndpoint = highEndpoint;
+            return getThis();
+        }
+
+        /**
+         * Calculate the key prefix length for the returned values. This will be used to derive the primary key used in
+         * the calculated continuation.
+         * @return the length of the key prefix
+         */
+        protected int calculatePrefixLength() {
+            int prefixLength = subspace.pack().length;
+            while ((prefixLength < lowBytes.length) &&
+                   (prefixLength < highBytes.length) &&
+                   (lowBytes[prefixLength] == highBytes[prefixLength])) {
+                prefixLength++;
+            }
+            return prefixLength;
+        }
+
+        public FDBRecordContext getContext() {
+            return context;
+        }
+
+        public int getLimit() {
+            return limit;
+        }
+
+        public int getPrefixLength() {
+            return prefixLength;
+        }
+
+        public ReadTransaction getTransaction() {
+            return transaction;
+        }
+
+        public CursorLimitManager getLimitManager() {
+            return limitManager;
+        }
+
+        public int getValuesLimit() {
+            return valuesLimit;
+        }
+
+        public boolean isReverse() {
+            return reverse;
+        }
+
+        public StreamingMode getStreamingMode() {
+            return streamingMode;
+        }
+
+        public KeySelector getBegin() {
+            return begin;
+        }
+
+        public KeySelector getEnd() {
+            return end;
+        }
+
+        private StreamingMode calcStreamingMode(final CursorStreamingMode propertiesStreamingMode, final int limit) {
+            if (propertiesStreamingMode == CursorStreamingMode.ITERATOR) {
+                return StreamingMode.ITERATOR;
+            } else if (propertiesStreamingMode == CursorStreamingMode.LARGE) {
+                return StreamingMode.LARGE;
+            } else if (propertiesStreamingMode == CursorStreamingMode.MEDIUM) {
+                return StreamingMode.MEDIUM;
+            } else if (propertiesStreamingMode == CursorStreamingMode.SMALL) {
+                return StreamingMode.SMALL;
+            } else if (limit == ReadTransaction.ROW_LIMIT_UNLIMITED) {
+                return StreamingMode.WANT_ALL;
+            } else {
+                return StreamingMode.EXACT;
+            }
+        }
+
+        protected abstract T getThis();
```

Comment:
minor: I believe I've seen this referred to as "self" and the type parameter as "self type" (for example, in this PEP: https://peps.python.org/pep-0673/). It may be useful to refer to it as such here

Reply:
Done.